src/client/datascience/kernel-launcher/localPythonAndRelatedNonPythonKernelSpecFinder.ts

```diff
@@ -0,0 +1,373 @@
+                        return false;
+                    }
+                    return true;
+                })
+                .map(async (k) => {
+                    // Find the interpreter that matches. If we find one, we want to use
+                    // this to start the kernel.
+                    const matchingInterpreter = this.findMatchingInterpreter(k, interpreters);
+                    if (matchingInterpreter) {
+                        const result: PythonKernelConnectionMetadata = {
+                            kind: 'startUsingPythonInterpreter',
+                            kernelSpec: k,
+                            interpreter: matchingInterpreter,
+                            id: getKernelId(k, matchingInterpreter)
+                        };
+
+                        // If interpreters were found, remove them from the interpreter list we'll eventually
+                        // return as interpreter only items
+                        filteredInterpreters = filteredInterpreters.filter((i) => matchingInterpreter !== i);
+
+                        // Return our metadata that uses an interpreter to start
+                        return result;
+                    } else {
+                        let interpreter = k.language === PYTHON_LANGUAGE ? activeInterpreter : undefined;
+                        // If the interpreter information is stored in kernelspec.json then use that to determine the interpreter.
+                        // This can happen under the following circumstances:
+                        // 1. Open workspace folder XYZ, and create a virtual environment named venvA
+                        // 2. Now assume we don't have raw kernels, and a kernel gets registered for venvA in kernelspecs folder.
+                        // 3. The kernel spec will contain metadata pointing to venvA.
+                        // 4. Now open a different folder (e.g. a sub directory of XYZ or a completely different folder).
+                        // 5. Now venvA will not be listed as an interpreter as Python will not discover this.
+                        // 6. However the kernel we registered against venvA will be in global kernels folder
+                        // In such an instance the interpreter information is stored in the kernelspec.json file.
+                        if (
+                            k.language === PYTHON_LANGUAGE &&
+                            k.metadata?.interpreter?.path &&
+                            k.metadata?.interpreter?.path !== activeInterpreter?.path
+                        ) {
+                            interpreter = await this.interpreterService
+                                .getInterpreterDetails(k.metadata?.interpreter?.path)
+                                .catch((ex) => {
+                                    traceError(`Failed to get interpreter details for Kernel Spec ${k.specFile}`, ex);
```

Comment:
Looks like we just proceed with the interpreter here, should we do that? I would think that if we failed to get details here then we don't have a valid interpreter, but I'm not fully sure if that's the case for getInterpreterDetails.

Reply:
This is old code, but yes, you are right.
Its nice to see this code is getting cleaned up (i.e. minor issues being picked up & resolved).

Reply:
@DonJayamanne Yeah, sorry for picking on old code here, I know that it's not stuff that you added. Just this review gives a chance to look at it more as you noted.